### DISCUSSION TITLE
Additional authenticator "Condition - request parameter"

### DISCUSSION THREAD

**Discussion by danielFesenmeyer** [keycloak/keycloak#21158] [Ideas]
In order to provide different authentication based on the login url, we have implemented a conditional authenticator, which decides based on a request parameter (actually, a request parameter specified in the initiating login request, which is kept by Keycloak as client note in the authentication session).

We "misused" that authenticator to implement "OTP reset" inside the Browser flow and post login flow - we have a somewhat "advanced" use case where users should only login via IDP, i.e. they don't have a Keycloak password, but need OTP - stored in Keycloak - for Keycloak login.
For this purpose, we checked whether the parameter "otp-reset=1" was specified: If yes, we triggered the "OTP reset", otherwise the standard flow was executed.

I'm quite sure that such a simple conditional authenticator can be useful in many use cases.

The configuration of the authenticator is currently just:

- **Parameter name:** _string_
- **Parameter value**: _string_
- **Negate output**: _boolean_, whether the check should be negated (i.e. the parameter does _not_ have the value, or does not even exist)

Let me know when you are interested - we are happy to contribute this authenticator.
When requested, we can also add more flexibility to the configuration, such as checking for existence of a parameter, using regex for value matching, ...

**Comment by slominskir** [comment]
Take a look at this discussion:  https://github.com/keycloak/keycloak/discussions/17152.   There is already a relatively new conditional flow framework for handling conditions, though it is complex and inflexible (and not documented well).   In practice, creating a custom authenticator that accepts request parameters and uses conditional logic often works best at the moment.  The OTP case is handled by the new condition flow I believe (and there is already an existing [ConditonalOTPFormAuthenticator](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/browser/ConditionalOtpFormAuthenticator.java) in Keycloak).   There are other scenarios that may not be though, like multiple forms with different user storage providers one after the other until one works.
